rfc/egress-billing.md

```diff
@@ -0,0 +1,309 @@
+`Origin`-gating and other CDN-esque tools to control their egress costs.
+
+```mermaid
+sequenceDiagram
+  Alice->>Gateway: Do you have bafk...7fi?
+  Gateway->>Rate Limit Service: Has bafk...7fi exceeded rate limits?
+  Rate Limit Service-->>Gateway: YES or NO or UNKNOWN
+  Gateway-)Accounting Service: Serving/not serving bafk...7fi
+  Gateway-->>Alice: 200 if YES or UNKNOWN, 429 if NO
+  Accounting Service->>Rate Limit Service: Update rate limits
+```
+
+## Introduction
+
+We currently operate the IPFS gateway at https://w3s.link as a free service. As a result, it is an unbound source of "cost"
+for our business. Covering some amount of the costs associated with this service make sense as a marketing and community
+building expense, but operating it for free makes it prohibitively expensive to give our customers acceptable read performance
+for their applications through our gateway.
+
+In order to fix this, we propose a design for billing our customers for data egress related to their applications. The suggested
+design allows us to maintain extremely fast read performance by allowing us to configure the "acceptable level" of "unpaid"
+data egress and giving our customers a way to authorize paid data egress using the "content claims" returned by the 
+[blob upload service].
+
+## Rate Limits
+
+Business realities (runway, need for marketing and community growth, etc) will determine an acceptable level of data egress
+through our gateway that can be paid for by the business. We will call this _R_, the "number of requests we're willing to pay for."
+
+We will use a tool like Redis to implement this rate limit, because it has a long history of being used for this and
+the right set of primitives to support it well: https://redis.io/glossary/rate-limiting/ - we will use Redis as an example
+in this document but recommend researching other tools like memcached and Caddy before making a final implementation decision.
+
+When the gateway receives a request for some content, it will make a request to Redis for the relevant rate limit data and decide to
+serve the request based on the response. In order to maintain acceptable read performance, we recommend avoiding any other network
+IO in making the decision to serve a piece of content. In the event that content is served as a result of the rate limit caching service
+being out of date, we count that request against _R_. 
+
+The following TypeScript pseudocode illustrates the rate limiting algorithm:
+
+
+```ts
+enum RateLimited {
+  Yes,
+  No,
+  Maybe,
+}
+
+async function handle(request){
+  const rateLimited = checkRedisToDetermineIfRateLimited(request)
+  const cid = getCidFromRequest(request)
+  switch (rateLimited) {
+    case Yes: 
+      return { status: 429, message: `${cid} is currently rate limited, please try again later` }
+    case No:
+      return { status: 200, body: readContentFromW3up(cid) }
+    case Maybe:
+      return { status: 200, body: readContentFromW3up(cid) }
```

Comment:
ah I was worried that people who aren't JS-brained would misread that as "if No, do nothing" - I could definitely be convinced but imho the way I've done it makes it a little clearer that something should be done in each case, and since they are right next to eachother it's pretty easy to verify that they're the same thing...